capsules/src/usb/ctap.rs

```diff
@@ -0,0 +1,401 @@
+                controller,
+                device_descriptor_buffer,
+                other_descriptor_buffer,
+                Some(&HID_DESCRIPTOR),
+                Some(&REPORT),
+                LANGUAGES,
+                strings,
+            ),
+            buffers: [Buffer64::default(), Buffer64::default()],
+            client: OptionalCell::empty(),
+            send_buffer: TakeCell::empty(),
+            recv_buffer: TakeCell::empty(),
+            recv_len: Cell::new(0),
+            recv_offset: Cell::new(0),
+            saved_endpoint: OptionalCell::empty(),
+        }
+    }
+
+    #[inline]
+    fn controller(&self) -> &'a U {
+        self.client_ctrl.controller()
+    }
+
+    #[inline]
+    fn buffer(&'a self, i: usize) -> &'a [VolatileCell<u8>; 64] {
+        &self.buffers[i - 1].buf
+    }
+
+    pub fn set_client(&'a self, client: &'a dyn hil::usb_hid::Client<'a>) {
+        self.client.set(client);
+    }
+
+    fn can_receive(&'a self) -> bool {
+        self.client
+            .map(move |client| client.can_receive())
+            .unwrap_or(false)
+    }
+}
+
+impl<'a, U: hil::usb::UsbController<'a>> hil::usb_hid::UsbHid<'a> for CtapHid<'a, U> {
+    fn send_buffer(
+        &'a self,
+        send: &'static mut [u8; 64],
+    ) -> Result<usize, (ReturnCode, &'static mut [u8; 64])> {
+        let len = send.len();
+
+        self.send_buffer.replace(send);
+        self.controller().endpoint_resume_in(ENDPOINT_IN_NUM);
+
+        Ok(len)
+    }
+
+    fn receive_buffer(
+        &'a self,
+        recv: &'static mut [u8; 64],
+    ) -> Result<(), (ReturnCode, &'static mut [u8; 64])> {
+        self.recv_buffer.replace(recv);
+
+        if self.saved_endpoint.is_some() {
+            // We have saved data from before, let's pass it.
+            if self.can_receive() {
+                self.recv_buffer.take().map(|buf| {
+                    self.client.map(move |client| {
+                        client.packet_received(
+                            buf,
+                            self.recv_offset.get(),
+                            self.saved_endpoint.take().unwrap(),
+                        );
+                    });
+                });
+                // Reset the offset
+                self.recv_offset.set(0);
+            }
+        } else {
+            // If we have nothing to process, accept more data
+            self.controller().endpoint_resume_out(ENDPOINT_OUT_NUM);
+        }
+
+        Ok(())
+    }
+
+    fn receive_cancel(&'a self) -> &'static mut [u8; 64] {
+        unimplemented!()
+    }
+}
+
+impl<'a, U: hil::usb::UsbController<'a>> hil::usb::Client<'a> for CtapHid<'a, U> {
+    fn enable(&'a self) {
+        // Set up the default control endpoint
+        self.client_ctrl.enable();
+
+        // Setup buffers for IN and OUT data transfer.
+        self.controller()
+            .endpoint_set_in_buffer(ENDPOINT_IN_NUM, self.buffer(ENDPOINT_IN_NUM));
+        self.controller()
```

Comment:
That's not the case though. I don't think that is even allowed in the USB spec.

The CTAP spec specifies two endpoints: https://fidoalliance.org/specs/fido-v2.0-id-20180227/fido-client-to-authenticator-protocol-v2.0-id-20180227.html#usb-descriptors

Reply:
See comment above, it's perfectly valid because the endpoint number is using 7 bits and the MSB is used to distinguish the direction.

Reply:
> That's not the case though. I don't think that is even allowed in the USB spec.

That's what we've implemented in OpenSK, and it works. Do you have an actual reference from the USB spec that forbids it?

Reply:
I have implemented two here and it also works.

Apparently a IN and OUT endpoint is allowed, sorry about that. In this case thought the CTAP spec still specifies two.

Reply:
I assumed that the specs requires one to use endpoint 1 in both modes, but indeed maybe it would work to pick (2, IN) and (5, OUT) if one wants to. However, I haven't tested any of that on a host application, so maybe it wouldn't be compatible with all host applications.

Reply:
Marking as resolved as this works with the Yubico tests and follows the CTAP spec (although there are also other interpretations of the spec). Feel free to unresolve it if you feel it's not addressed.